Todo/AppDelegate.swift

```diff
@@ -0,0 +1,420 @@
+
+            if kEncryptionEnabled {
+                if let encryptionKey = key {
+                    options.encryptionKey = encryptionKey
+                }
+            }
+
+            try database = CBLManager.sharedInstance().openDatabaseNamed(dbname, withOptions: options)
+            if newKey != nil {
+                try database.changeEncryptionKey(newKey)
+            }
+            startConflictLiveQuery()
+    }
+    
+    func closeDatabase() throws {
+        stopConflictLiveQuery()
+        try database.close()
+    }
+
+    // MARK: - Login
+
+    func login(username: String? = nil) {
+        let storyboard =  window!.rootViewController!.storyboard
+        let navigation = storyboard!.instantiateViewControllerWithIdentifier(
+            "LoginNavigationController") as! UINavigationController
+        let loginController = navigation.topViewController as! LoginViewController
+        loginController.delegate = self
+        loginController.username = username
+        window!.rootViewController = navigation
+    }
+
+    func logout() {
+        stopReplication()
+        do {
+            try closeDatabase()
+        } catch let error as NSError {
+            NSLog("Cannot close database: %@", error)
+        }
+        let oldUsername = Session.username
+        Session.username = nil
+        login(oldUsername)
+    }
+
+    func showApp() {
+        guard let root = window?.rootViewController, storyboard = root.storyboard else {
+            return
+        }
+
+        let controller = storyboard.instantiateInitialViewController()
+        window!.rootViewController = controller
+    }
+
+    // MARK: - LoginViewControllerDelegate
+
+    func login(controller: UIViewController, withUsername username: String,
+        andPassword password: String) {
+            processLogin(controller, withUsername: username, withPassword: password)
+    }
+
+    func processLogin(controller: UIViewController, withUsername username: String,
+        withPassword password: String, withNewPassword newPassword: String? = nil) {
+            do {
+                try startSession(withUsername: username, withPassword: password,
+                    withNewPassword: newPassword)
+            } catch let error as NSError {
+                if error.code == 401 {
+                    handleEncryptionError(controller, withUsername: username,
+                        withPassword: password)
+                } else {
+                    Ui.showMessageDialog(
+                        onController: controller,
+                        withTitle: "Error",
+                        withMessage: "Login has an error occurred, code = \(error.code).")
+                    NSLog("Cannot start a session: %@", error)
+                }
+            }
+    }
+
+    func handleEncryptionError(controller: UIViewController, withUsername username: String,
+        withPassword password: String) {
+            Ui.showTextInputDialog(
```

Comment:
Agree. It's really wrong. The method is generic used in the some other parts of the code with different configuration. In Ui.swift, I could create a specific method for showing the encryption rekey dialog so it will be much shorter in here.